enhancements/security/selinux-operator.md

```diff
@@ -0,0 +1,269 @@
+
+#### Security (RBAC)
+
+As a system administrator, I want to enforce that only certain users can create
+policies in certain namespaces.
+
+This way, only trusted users would be able to install policies in the system.
+
+#### Security (namespace separation)
+
+As a system administrator, I want to make sure that workloads can only use
+approved policies.
+
+This way, an attacker can't take into use a policy from a namespace that's not
+accessible.
+
+### API Specification
+
+This operator would expose a CRD that would look as follows:
+
+```
+apiVersion: selinux.openshift.io/v1alpha1
+kind: SelinuxPolicy
+metadata:
+  name: virt-launcher
+  namespace: my-namespace
+spec:
+  apply: true
+  policy: |
+    (blockinherit container)
+    (typeattributeset sandbox_net_domain (process))
+    (allow process kernel_t (system (module_request)))
+    (allow process mtrr_device_t (file (write)))
+    (allow process self (tun_socket (relabelfrom)))
+    (allow process self (tun_socket (relabelto)))
+    (allow process self (tun_socket (attach_queue)))
+    (allow process sysfs_t (file (write)))
+    (allow process tmp_t (dir (write add_name open getattr setattr read link search remove_name reparent lock ioctl)))
+    (allow process tmp_t (file (setattr open read write create getattr append ioctl lock)))
+    (allow process container_share_t (dir (write add_name)))
+    (allow process container_share_t (file (create setattr write )))
+    (allow process container_var_run_t (dir (write add_name open getattr setattr read link search remove_name reparent lock ioctl)))
+    (allow process container_var_run_t (file (setattr open read write create getattr append ioctl lock)))
+status:
+  usage: virt-launcher_my-namespace.process
+  state: INSTALLED
+```
+
+Where:
+
+* `spec.apply`: Is a flag that allows us to tell the operator if the policy is to be
+  applied (installed) or not. Having this flag allows policy writers and system
+  administrators to verify the policy before taking it into use in the system.
+  Automated tools such as the aforementioned `selinux-policy-helper-operator`
+  would generate policies with this flag set to `false`.
+* `spec.policy`: Contains the actual lines of policy in
+  [CIL language](https://github.com/SELinuxProject/cil/wiki). The decision for
+  this language is because `Udica` already writes policies in such a language,
+  and it's an easier to read format than the typical SELinux type enforcement
+  (te) language.
+* `status.usage`: This string indicates how the user is supposed to use this
+  policy in their `Pod` templates. The format is as follows:
+  `<policy name>_<policy namespace>.process>`
+* `status.state`: This represents the current status of the policy in the
+  cluster. The accepted states are: `PENDING`, `IN-PROGRESS`, `INSTALLED`,
+  `ERROR`. A typical reason why a policy would end up in an ERROR state would
+  be that it has errors in the syntax.
+
+
+### Implementation Details/Notes/Constraints
```

Comment:
Is there a reason that the machine config operator is not the desired vehicle to distribute this configuration?

Reply:
There are several reasons:

One of them is that SELinux policies are not just files that we persist on the host and they get used (as would be the case with Seccomp), there needs to be further automation and tracking of these (gotta trigger an installation command, and if you want to remove it, gotta trigger an uninstall command too).

On the other hand, this proposal is beyond just installing the policies in a node. The idea is to namespace them and is able to enforce RBAC on them (some policies might be fairly harmless while others might give the containers a lot of privilege), so we want to be careful who can create these.

Also, creating a MachineConfig triggers a node restart, which would not be ideal if you want to take a policy into use immediately. Let's look at the following example: We want to install an operator from OperatorHub. This operator has been hardened and it comes with a read-to-use SELinux policy for its operand. We just want to install it and use it immediately. If it would use MachineConfig to distribute the policy it would trigger a restart, and if you at some point decide to uninstall the operator (cause you have a different solution somewhere) this makes it really hard to track what policies do you have in your system. So you really want a way to be able to track these and remove them when they're no longer needed.

Finally, UX. SELinux is normally seen as a very complicated technology. This aims to simplify things by exposing it as an object, as opposed to having to decode a url-encoded string from a MachineConfig. There is also a helper operator that automatically generates policies if you request so https://github.com/JAORMX/selinux-policy-helper-operator.

Reply:
OK, wrote this into the `alternatives` section